# U2A1. Operaciones Matriciales y Vectoriales  

## Elaborado por: Gabriel Carrizales  

### Fecha: 22/09/2025  

#### Materia: Robótica  

##### Docente: ING. Osbaldo Aragón Banderas 


# Introducción

El álgebra lineal es una herramienta fundamental en la ingeniería y la programación científica, ya que permite modelar y resolver problemas relacionados con matrices, vectores y sistemas lineales. Esta práctica tiene como objetivo aplicar los conceptos teóricos de operaciones matriciales y vectoriales mediante la implementación de códigos en Python, utilizando Jupyter Notebook como entorno de desarrollo.

A través de los ejercicios propuestos, se busca reforzar habilidades en la manipulación de matrices, cálculo de determinantes, inversión de matrices, resolución de sistemas de ecuaciones, y operaciones vectoriales como producto cruz, producto escalar y proyecciones. Asimismo, se incluyen técnicas avanzadas como la ortogonalización de Gram–Schmidt y el cálculo del espacio nulo de una matriz.

El desarrollo de esta práctica permite a los estudiantes comprender cómo los conceptos de álgebra lineal se aplican en la programación, promoviendo un aprendizaje integral que combina teoría matemática y habilidades computacionales.

Este notebook contiene el desarrollo de **10 ejercicios** de álgebra lineal utilizando Python, con el apoyo de las librerías `numpy` y `sympy`.  



A lo largo de este notebook se realizan las siguientes actividades:

1. **Suma de matrices**  
2. **Multiplicación de matrices**  
3. **Inversión de matriz**  
4. **Resolución de sistemas de ecuaciones lineales**  
5. **Cálculo de determinante**  
6. **Producto cruz de vectores**  
7. **Proyección ortogonal**  
8. **Producto escalar de proyecciones**  
9. **Ortogonalización de Gram–Schmidt**  
10. **Cálculo de espacio nulo**

Cada ejercicio incluye:
- Una breve explicación del problema.  
- El código necesario para resolverlo.  
- La salida correspondiente.

De esta manera, el notebook no solo muestra los resultados, sino que también sirve como **guía práctica** para aprender a implementar operaciones de álgebra lineal en Python.  


## Objetivo General  
Aplicar los conceptos de álgebra lineal mediante el uso de Python en Jupyter Notebook, resolviendo problemas de operaciones matriciales y vectoriales que fortalezcan la comprensión teórica y la habilidad práctica en programación científica.  

## Objetivos Específicos  
- Implementar operaciones básicas de matrices como suma, multiplicación, determinante e inversa.  
- Resolver sistemas de ecuaciones lineales utilizando métodos matriciales.  
- Calcular productos vectoriales como el producto cruz y el producto escalar.  
- Determinar proyecciones ortogonales de vectores en distintos subespacios.  
- Aplicar el proceso de ortogonalización de Gram–Schmidt para generar conjuntos ortogonales.  
- Identificar la base del espacio nulo de una matriz a partir de su forma reducida.  
- Desarrollar habilidades en el uso de librerías matemáticas de Python (`numpy` y `sympy`).  
- Fortalecer la conexión entre teoría matemática y aplicaciones computacionales en ingeniería.  


In [3]:
# Celda 0: Imports y configuración básica
# Importación de librerías necesarias para operaciones matriciales y vectoriales
import numpy as np
from sympy import Matrix, Rational

# Configuración de impresión de numpy:
# - precision=8 para mostrar hasta 8 decimales
# - suppress=True para evitar notación científica
np.set_printoptions(precision=8, suppress=True)


# Ejercicio 1: Suma de matrices
Objetivo: Sumar las matrices A y B y mostrar la matriz resultante C = A + B.


In [4]:
# Ejercicio 1 - Suma de matrices
# Definición de las matrices A y B
A = np.array([[2, 4, 6],
              [1, 3, 5],
              [7, 9, 11]])
B = np.array([[12, 10, 8],
              [6, 4, 2],
              [0, -2, -4]])

# Suma de matrices
C = A + B

# Impresión de resultados de forma clara y ordenada
print("Matriz A:")
print(A, "\n")
print("Matriz B:")
print(B, "\n")
print("Matriz resultante C = A + B:")
print(C)


Matriz A:
[[ 2  4  6]
 [ 1  3  5]
 [ 7  9 11]] 

Matriz B:
[[12 10  8]
 [ 6  4  2]
 [ 0 -2 -4]] 

Matriz resultante C = A + B:
[[14 14 14]
 [ 7  7  7]
 [ 7  7  7]]


# Ejercicio 2: Multiplicación de matrices
Objetivo: Multiplicar A * B donde A es 3x2 y B es 2x2. Mostrar la matriz resultante C.


In [5]:
# Ejercicio 2 - Multiplicación de matrices
# Definición de las matrices A y B
A = np.array([[2, 1],
              [3, 4],
              [5, 6]])
B = np.array([[7, 8],
              [9, 10]])

# Multiplicación de matrices usando el operador @ (o np.dot)
C = A @ B

# Impresión de resultados de forma clara y ordenada
print("Matriz A:")
print(A, "\n")
print("Matriz B:")
print(B, "\n")
print("Matriz resultante C = A * B:")
print(C)


Matriz A:
[[2 1]
 [3 4]
 [5 6]] 

Matriz B:
[[ 7  8]
 [ 9 10]] 

Matriz resultante C = A * B:
[[ 23  26]
 [ 57  64]
 [ 89 100]]


# Ejercicio 3: Inversión de matriz
Objetivo: Calcular la inversa de la matriz cuadrada A. Mostraremos una versión numérica (float) y la inversa exacta con fracciones usando sympy.


In [6]:
# Ejercicio 3 - Inversa de A (numérica y exacta)
# Definición de la matriz A
A = np.array([[4, 7, 2],
              [2, 6, 8],
              [3, 1, 9]])

# Cálculo de la inversa numérica utilizando numpy (resultados en float)
A_inv_num = np.linalg.inv(A)
print("Matriz A:")
print(A, "\n")
print("Inversa de A (numpy, decimal):")
print(A_inv_num, "\n")

# Cálculo de la inversa exacta utilizando sympy (resultados en fracciones)
A_sym = Matrix([[4,7,2],
                [2,6,8],
                [3,1,9]])
A_inv_exact = A_sym.inv()
print("Inversa de A (sympy, exacta en fracciones):")
print(A_inv_exact)


Matriz A:
[[4 7 2]
 [2 6 8]
 [3 1 9]] 

Inversa de A (numpy, decimal):
[[ 0.2371134  -0.31443299  0.22680412]
 [ 0.03092784  0.15463918 -0.1443299 ]
 [-0.08247423  0.08762887  0.05154639]] 

Inversa de A (sympy, exacta en fracciones):
Matrix([[23/97, -61/194, 22/97], [3/97, 15/97, -14/97], [-8/97, 17/194, 5/97]])


# Ejercicio 4: Resolución de sistema de ecuaciones lineales
Sistema:
2x + y + z = 8
3x + 5y + 2z = 21
x + 2y + 4z = 11

Se escribe AX = B y se resuelve para X = [x, y, z]. Mostraremos solución numérica y exacta.


In [7]:
# Ejercicio 4 - Resolución del sistema AX = B
# Definición de la matriz de coeficientes A y el vector de términos B
A = np.array([[2, 1, 1],
              [3, 5, 2],
              [1, 2, 4]])
B = np.array([8, 21, 11])

# Resolución numérica usando numpy (resultado en decimales)
X = np.linalg.solve(A, B)
print("Sistema de ecuaciones AX = B")
print("Matriz A:")
print(A, "\n")
print("Vector B:")
print(B, "\n")
print("Solución numérica X:")
print(X, "\n")

# Resolución exacta usando sympy (resultados en fracciones)
A_sym = Matrix([[2,1,1],
                [3,5,2],
                [1,2,4]])
B_sym = Matrix([8,21,11])
X_sym = A_sym.LUsolve(B_sym)
print("Solución exacta X (fracciones):")
print(X_sym)


Sistema de ecuaciones AX = B
Matriz A:
[[2 1 1]
 [3 5 2]
 [1 2 4]] 

Vector B:
[ 8 21 11] 

Solución numérica X:
[2.30434783 2.43478261 0.95652174] 

Solución exacta X (fracciones):
Matrix([[53/23], [56/23], [22/23]])


# Ejercicio 5: Cálculo de determinante
Objetivo: Calcular el determinante de la matriz A.


In [8]:
# Ejercicio 5 - Cálculo del determinante
# Definición de la matriz A
A = np.array([[3, -2, 1],
              [0, 5, 4],
              [2, 1, 7]])

# Cálculo del determinante numérico usando numpy
det_A = np.linalg.det(A)
print("Matriz A:")
print(A, "\n")
print("Determinante (numpy, decimal):", det_A, "\n")

# Cálculo del determinante exacto usando sympy (fracciones)
A_sym = Matrix([[3,-2,1],
                [0,5,4],
                [2,1,7]])
det_exact = A_sym.det()
print("Determinante exacto (sympy):", det_exact)


Matriz A:
[[ 3 -2  1]
 [ 0  5  4]
 [ 2  1  7]] 

Determinante (numpy, decimal): 67.00000000000004 

Determinante exacto (sympy): 67


# Ejercicio 6: Producto cruz de vectores
Dados A y B en R^3, calcular A x B.


In [9]:
# Ejercicio 6 - Producto cruz de vectores
# Definición de los vectores A y B en R^3
A = np.array([2, 3, -1])
B = np.array([1, -2, 4])

# Cálculo del producto cruz utilizando numpy
cross = np.cross(A, B)

# Impresión de resultados de forma clara
print("Vector A:")
print(A, "\n")
print("Vector B:")
print(B, "\n")
print("Producto cruz A x B:")
print(cross)


Vector A:
[ 2  3 -1] 

Vector B:
[ 1 -2  4] 

Producto cruz A x B:
[10 -9 -7]


# Ejercicio 7: Proyección ortogonal
Calcular la proyección de V sobre U: proj_U(V) = (V·U / U·U) * U


In [10]:
# Ejercicio 7 - Proyección ortogonal de V sobre U
# Definición de los vectores V y U
V = np.array([5, -3, 2])
U = np.array([2, 1, 2])

# Cálculo del factor de proyección y la proyección usando numpy
factor = np.dot(V, U) / np.dot(U, U)
proj_V_on_U = factor * U

# Impresión de resultados numéricos
print("Vector V:", V)
print("Vector U:", U)
print("Factor de proyección (V·U / U·U) =", factor)
print("Proyección de V sobre U (numpy) =", proj_V_on_U, "\n")

# Cálculo exacto usando sympy (fracciones)
V_sym = Matrix([5,-3,2])
U_sym = Matrix([2,1,2])
proj_exact = (V_sym.dot(U_sym) / U_sym.dot(U_sym)) * U_sym

# Impresión de la proyección exacta
print("Proyección exacta de V sobre U (sympy):", proj_exact)


Vector V: [ 5 -3  2]
Vector U: [2 1 2]
Factor de proyección (V·U / U·U) = 1.2222222222222223
Proyección de V sobre U (numpy) = [2.44444444 1.22222222 2.44444444] 

Proyección exacta de V sobre U (sympy): Matrix([[22/9], [11/9], [22/9]])


# Ejercicio 8: Producto escalar de proyecciones
Dados V, U, W, calcule el producto escalar entre proj_U(V) y proj_W(V).


In [11]:
# Ejercicio 8 - Producto escalar de proyecciones
# Definición de los vectores V, U y W
V = np.array([3, -1, 2])
U = np.array([2, 2, -1])
W = np.array([1, 4, -2])

# Cálculo de las proyecciones de V sobre U y W usando numpy
proj_VU = (np.dot(V, U) / np.dot(U, U)) * U
proj_VW = (np.dot(V, W) / np.dot(W, W)) * W

# Producto escalar entre las proyecciones
dot_proj = np.dot(proj_VU, proj_VW)

# Impresión de resultados numéricos
print("Proyección de V sobre U (numpy):", proj_VU)
print("Proyección de V sobre W (numpy):", proj_VW)
print("Producto escalar entre las proyecciones (numpy):", dot_proj, "\n")

# Cálculo exacto usando sympy
V_sym = Matrix([3,-1,2])
U_sym = Matrix([2,2,-1])
W_sym = Matrix([1,4,-2])

proj_VU_sym = (V_sym.dot(U_sym) / U_sym.dot(U_sym)) * U_sym
proj_VW_sym = (V_sym.dot(W_sym) / W_sym.dot(W_sym)) * W_sym
dot_exact = proj_VU_sym.dot(proj_VW_sym)

# Impresión del producto escalar exacto
print("Producto escalar exacto entre las proyecciones (sympy):", dot_exact)



Proyección de V sobre U (numpy): [ 0.44444444  0.44444444 -0.22222222]
Proyección de V sobre W (numpy): [-0.23809524 -0.95238095  0.47619048]
Producto escalar entre las proyecciones (numpy): -0.6349206349206349 

Producto escalar exacto entre las proyecciones (sympy): -40/63


# Ejercicio 9: Ortogonalización de Gram–Schmidt
Dado el conjunto de vectores v1, v2, v3, aplicar Gram–Schmidt para obtener un conjunto ortogonal.
(No normalizamos, solo obtenemos vectores ortogonales).


In [13]:
# Ejercicio 9 - Gram-Schmidt
def gram_schmidt(vectors):
    ortogonales = []
    for v in vectors:
        w = v.astype(float).copy()
        for u in ortogonales:
            proj = (np.dot(v, u) / np.dot(u, u)) * u
            w = w - proj
        ortogonales.append(w)
    return ortogonales

v1 = np.array([1,1,0], dtype=float)
v2 = np.array([1,2,1], dtype=float)
v3 = np.array([2,1,3], dtype=float)

ort = gram_schmidt([v1, v2, v3])
for i, vec in enumerate(ort, 1):
    print(f"u{i} (ortogonal) =", vec)

# resultado exacto (fracciones) con sympy
v1s = Matrix([1,1,0])
v2s = Matrix([1,2,1])
v3s = Matrix([2,1,3])
def gs_sympy(vs):
    ort = []
    for v in vs:
        w = v
        for u in ort:
            w = w - (v.dot(u)/u.dot(u))*u
        ort.append(w)
    return ort

ort_sym = gs_sympy([v1s, v2s, v3s])
print("\nOrtogonales exactos (sympy):")
for vec in ort_sym:
    print(vec)


u1 (ortogonal) = [1. 1. 0.]
u2 (ortogonal) = [-0.5  0.5  1. ]
u3 (ortogonal) = [ 1.33333333 -1.33333333  1.33333333]

Ortogonales exactos (sympy):
Matrix([[1], [1], [0]])
Matrix([[-1/2], [1/2], [1]])
Matrix([[4/3], [-4/3], [4/3]])


# Ejercicio 10: Espacio nulo
Dada la matriz en forma escalonada reducida por filas, encuentra una base para el espacio nulo (soluciones de A v = 0).


In [13]:
# Ejercicio 10 - Espacio nulo (nullspace) usando sympy
# Definición de la matriz A en forma escalonada
A = Matrix([[1, 2, 0, 3],
            [0, 1, 0, 2],
            [0, 0, 1, 1]])

# Cálculo de la base del espacio nulo
null_space = A.nullspace()

# Impresión de los vectores que forman la base del espacio nulo
print("Base del espacio nulo (cada vector columna representa un vector base):")
for v in null_space:
    print(v)

# Nota: la solución general del sistema homogéneo Av = 0 se puede expresar
# como combinación lineal de estos vectores, es decir, t * (vector) donde t es un escalar.



Base del espacio nulo (cada vector columna representa un vector base):
Matrix([[1], [-2], [-1], [1]])


# Respuestas analíticas

**Ejercicio 1 — Suma de matrices**
A + B = C, donde

A = $\begin{pmatrix}2 & 4 & 6\\[4pt]1 & 3 & 5\\[4pt]7 & 9 & 11\end{pmatrix}$,
B = $\begin{pmatrix}12 & 10 & 8\\[4pt]6 & 4 & 2\\[4pt]0 & -2 & -4\end{pmatrix}$

$$
C = A + B = 
\begin{pmatrix}
14 & 14 & 14\\[4pt]
7  & 7  & 7\\[4pt]
7  & 7  & 7
\end{pmatrix}
$$

---

**Ejercicio 2 — Multiplicación de matrices**
A (3×2) · B (2×2) = C (3×2), con

A = $\begin{pmatrix}2 & 1\\[4pt]3 & 4\\[4pt]5 & 6\end{pmatrix}$,
B = $\begin{pmatrix}7 & 8\\[4pt]9 & 10\end{pmatrix}$

$$
C = A\cdot B =
\begin{pmatrix}
23 & 26\\[4pt]
57 & 64\\[4pt]
89 & 100
\end{pmatrix}
$$

---

**Ejercicio 3 — Inversa de la matriz A**
A = $\begin{pmatrix}4 & 7 & 2\\[4pt]2 & 6 & 8\\[4pt]3 & 1 & 9\end{pmatrix}$

La inversa exacta (entrada a entrada en fracciones) es:

$$
A^{-1} =
\begin{pmatrix}
\frac{23}{97} & -\frac{61}{194} & \frac{22}{97}\\[6pt]
\frac{3}{97}  & \frac{15}{97}   & -\frac{14}{97}\\[6pt]
-\frac{8}{97} & \frac{17}{194}  & \frac{5}{97}
\end{pmatrix}
$$

(Se puede multiplicar numéricamente para obtener aproximaciones decimales si lo prefieres.)

---

**Ejercicio 4 — Resolución del sistema lineal**
Sistema:

$$
\begin{cases}
2x + y + z = 8\\[4pt]
3x + 5y + 2z = 21\\[4pt]
x + 2y + 4z = 11
\end{cases}
$$

En forma matricial $AX=B$. La solución exacta (fracciones) es:

$$
X = 
\begin{pmatrix}
x\\[4pt] y\\[4pt] z
\end{pmatrix}
=
\begin{pmatrix}
\frac{53}{23}\\[4pt]
\frac{56}{23}\\[4pt]
\frac{22}{23}
\end{pmatrix}
$$

(siendo en decimal: $x\approx 2.3043,\; y\approx 2.4348,\; z\approx 0.9565$.)

---

**Ejercicio 5 — Determinante**
A = $\begin{pmatrix}3 & -2 & 1\\[4pt]0 & 5 & 4\\[4pt]2 & 1 & 7\end{pmatrix}$

$$
\det(A) = 67
$$

---

**Ejercicio 6 — Producto cruz de vectores**
A = $[2,\,3,\,-1]$, B = $[1,\,-2,\,4]$

$$
A \times B = 
\begin{pmatrix}10\\[4pt]-9\\[4pt]-7\end{pmatrix}
$$

---

**Ejercicio 7 — Proyección ortogonal de V sobre U**
V = $[5,\,-3,\,2]$, U = $[2,\,1,\,2]$

$$
\operatorname{proj}_U(V) = \left(\frac{V\cdot U}{U\cdot U}\right) U
= \frac{22}{9}\begin{pmatrix}1\\[4pt]\tfrac{1}{2}\\[4pt]1\end{pmatrix}
=
\begin{pmatrix}\tfrac{22}{9}\\[6pt]\tfrac{11}{9}\\[6pt]\tfrac{22}{9}\end{pmatrix}
$$

(aprox.: $[2.4444,\;1.2222,\;2.4444]$.)

---

**Ejercicio 8 — Producto escalar de proyecciones**
V = $[3,\,-1,\,2]$, U = $[2,\,2,\,-1]$, W = $[1,\,4,\,-2]$

Primero las proyecciones:

$$
\operatorname{proj}_{U}(V)=
\begin{pmatrix}\tfrac{4}{9}\\[4pt]\tfrac{4}{9}\\[4pt]-\tfrac{2}{9}\end{pmatrix},\qquad
\operatorname{proj}_{W}(V)=
\begin{pmatrix}-\tfrac{5}{21}\\[4pt]-\tfrac{20}{21}\\[4pt]\tfrac{10}{21}\end{pmatrix}
$$

El producto escalar entre ambas proyecciones es:

$$
\operatorname{proj}_U(V)\cdot\operatorname{proj}_W(V) = -\frac{40}{63}
\approx -0.63492
$$

---

**Ejercicio 9 — Ortogonalización (Gram–Schmidt)**
Vectores originales:
$v_1=[1,1,0],\; v_2=[1,2,1],\; v_3=[2,1,3]$.

Aplicando Gram–Schmidt (sin normalizar) se obtiene el conjunto ortogonal:

$$
u_1 = \begin{pmatrix}1\\[4pt]1\\[4pt]0\end{pmatrix},\qquad
u_2 = \begin{pmatrix}-\tfrac{1}{2}\\[6pt]\tfrac{1}{2}\\[6pt]1\end{pmatrix},\qquad
u_3 = \begin{pmatrix}\tfrac{4}{3}\\[6pt]-\tfrac{4}{3}\\[6pt]\tfrac{4}{3}\end{pmatrix}
$$

Puedes verificar que $u_i\cdot u_j = 0$ para $i\neq j$.

---

**Ejercicio 10 — Espacio nulo (nullspace)**
Matriz en RREF:

$$
A = \begin{pmatrix}
1 & 2 & 0 & 3\\[4pt]
0 & 1 & 0 & 2\\[4pt]
0 & 0 & 1 & 1
\end{pmatrix}
$$

Base del espacio nulo (cada vector columna es un generador):

$$
\text{Null}(A) = \operatorname{span}\left\{
\begin{pmatrix}1\\[4pt]-2\\[4pt]-1\\[4pt]1\end{pmatrix}
\right\}
$$

Es decir, la solución general de $Av=0$ es $v = t\begin{pmatrix}1\\-2\\-1\\1\end{pmatrix}$, $t\in\mathbb{R}$.





# Conclusión  

A lo largo de esta práctica se aplicaron distintos conceptos de álgebra lineal mediante ejercicios resueltos en Python dentro de Jupyter Notebook, reforzando la comprensión de operaciones con matrices y vectores. Se comprobó cómo las operaciones básicas, como suma, multiplicación, determinantes e inversas, son fundamentales para resolver problemas más complejos como sistemas de ecuaciones, proyecciones o el cálculo de espacios nulos.  

Además, se evidenció la utilidad de métodos como la ortogonalización de Gram–Schmidt, que permite construir bases ortogonales para facilitar cálculos posteriores, así como la importancia de interpretar de manera analítica y numérica cada resultado.  

En conclusión, esta práctica no solo fortaleció el dominio de las técnicas algebraicas, sino que también mostró la relevancia de su aplicación en programación y simulación, herramientas indispensables en la ingeniería para modelar y resolver problemas reales de manera eficiente y precisa.  
